crates/dbsp/src/circuit/circuit_builder.rs

```diff
@@ -1506,6 +1511,116 @@ where
     }
 }
 
+// TODO: use a better type thank serde_json::Value.
+#[derive(Default, Debug, Clone, Serialize, Deserialize)]
+pub struct CircuitMetadata {
+    metadata: HashMap<NodeId, serde_json::Value>,
+}
+
+#[derive(Default, Debug)]
+pub struct MetadataExchangeInner {
+    /// Metadata registered by operators running in the local circuit.
+    local_metadata: RefCell<CircuitMetadata>,
+
+    /// Metadata received from peers. All workers have identical metadata snapshots during a step.
+    global_metadata: RefCell<Vec<CircuitMetadata>>,
+}
+
+/// Metadata exchange.
+///
+/// Allows the circuit to exchange arbitrary semi-structured data with its peers.
+///
+/// Every operator in the circuit can update its local metadata.
+/// Before every step, the circuit broadcasts its metadata to all other workers
+/// and receives their metadata. As a result all workers have identical metadata
+/// snapshots during the step and can make deterministic decisions based on it, such
+/// as choosing a balancing policy for a stream.
+#[derive(Default, Debug, Clone)]
+pub struct MetadataExchange {
+    inner: Rc<MetadataExchangeInner>,
+}
+
+impl MetadataExchange {
+    fn new() -> Self {
+        Self::default()
+    }
+
+    /// Get the current snapshot of the local metadata registered by operators running in the local circuit.
+    pub fn local_metadata(&self) -> CircuitMetadata {
+        self.inner.local_metadata.borrow().clone()
+    }
+
+    /// Update the local metadata for the operator with the given id.
+    pub fn set_local_operator_metadata(&self, id: NodeId, metadata: serde_json::Value) {
+        self.inner
+            .local_metadata
+            .borrow_mut()
+            .metadata
+            .insert(id, metadata.clone());
+    }
+
+    /// Update the local metadata for the operator with the given id by serializing `metadata` to a JSON value.
+    pub fn set_local_operator_metadata_typed<T>(&self, id: NodeId, metadata: T)
+    where
+        T: Serialize,
+    {
+        self.inner
+            .local_metadata
+            .borrow_mut()
+            .metadata
+            .insert(id, serde_json::to_value(metadata).unwrap());
+    }
+
+    /// Get the current snapshot of the local metadata for the operator with the given id.
+    pub fn get_local_operator_metadata(&self, id: NodeId) -> Option<serde_json::Value> {
+        self.inner
+            .local_metadata
+            .borrow()
+            .metadata
+            .get(&id)
+            .cloned()
+    }
+
+    /// Set the global metadata received from peers (invoked by the scheduler).
```

Comment:
this data structure will be quadratic in the number of workers. I wonder whether it could be a problem with scale-out which would warrant a hierarchical representation (perhaps one copy per machine). Ideally you can design an API which hides this implementation detail